packages/integration_test/lib/integration_test_driver_extended.dart

```diff
@@ -6,12 +6,51 @@
 // ignore_for_file: avoid_print
 
 import 'dart:async';
+import 'dart:convert';
 import 'dart:io';
 
 import 'package:flutter_driver/flutter_driver.dart';
+import 'package:path/path.dart' as path;
 
 import 'common.dart';
 
+/// Flutter Driver test output directory.
+///
+/// Tests should write any output files to this directory. Defaults to the path
+/// set in the FLUTTER_TEST_OUTPUTS_DIR environment variable, or `build` if
+/// unset.
+String testOutputsDirectory =
+    Platform.environment['FLUTTER_TEST_OUTPUTS_DIR'] ?? 'build';
+
+/// The callback type to handle [Response.data] after the test
+/// succeeds.
+typedef ResponseDataCallback = FutureOr<void> Function(Map<String, dynamic>?);
+
+/// Writes a json-serializable data to
+/// [testOutputsDirectory]/`testOutputFilename.json`.
+///
+/// This is the default `responseDataCallback` in [integrationDriver].
+Future<void> writeResponseData(
+  Map<String, dynamic>? data, {
+  String testOutputFilename = 'integration_response_data',
+  String? destinationDirectory,
+}) async {
+  destinationDirectory ??= testOutputsDirectory;
+  await fs.directory(destinationDirectory).create(recursive: true);
+  final File file = fs.file(path.join(
+    destinationDirectory,
+    '$testOutputFilename.json',
+  ));
+  final String resultString = _encodeJson(data, true);
+  await file.writeAsString(resultString);
+}
+
+/// The exit function type for io.exit
+typedef ExitFn = Function(int code);
+
+/// the exit function to help test inject mock exit function
+ExitFn exitFn = exit;
```

Comment:
What is this doing here? Did I miss this in my previous review or was this added later? This doesn't look like API we should have and it also doesn't follow flutter style.

Reply:
Hi @goderbauer I guess you missed it in your previous review. This is for testing in `integration_test_driver_extended_test.dart`, as we call exit(0) and exit(1) in ` integrationDriver`, so have to find a way to mock `exit` function in test. The idea come from https://stackoverflow.com/questions/32990067/dart-how-do-you-write-a-test-against-a-function-that-calls-exit. Do you have a better solution to mock it?

Reply:
Rather than mocking the `exit` call, I think we'd be better off doing something like starting a Process that runs scripts for success or failure states in the test and checking the exit code of that process.

We do something similar in `binding_fail_test.dart`, though it doesn't make use of the process to check the exit code there. I see two ways forward here:

* If you'd like to validate the exit call specifically, see if you can make use of a Process as described above.
* Since there's not currently precedent for testing the exit call for either of the drivers, I think it's fair to exclude the exit validation for the moment. We can log an issue to add in that validation separately from this PR.

Reply:
Hi @pdblasi-google, I have made some revisions to this PR. I have used the Process class to encapsulate the test in order to prevent the exit function from interrupting the test. However, I am still unable to verify the exit code for it, since the exit code pertains to flutter drive and not flutter test. If I use the Process to call flutter drive, the unit test will essentially become an integration test. I am unsure if the benefits outweigh the costs in this case. Therefore, the unit test only checks different scenarios for calling the callback function, without verifying the exit function. I would appreciate your input on this matter.

Reply:
Does `process.exitCode` in your new tests not return the code passed to the `exit` method in the `integrationDriver`?

Reply:
"No, the process.exitCode is always 1 because it is related to the Process.run('flutter', 'test', 'xxxxxx.dart') and not to the integrationDriver. Therefore, the exit method in integrationDriver interrupts the execution of Process.run('flutter', 'test', 'xxxxx.dart'), resulting in an exit code of 1."

Reply:
No, the process.exitCode is always 1 because it is related to the Process.run('flutter', 'test', 'xxxxxx.dart') and not to the integrationDriver. Therefore, the exit method in integrationDriver interrupts the execution of Process.run('flutter', 'test', 'xxxxx.dart'), resulting in an exit code of 1.

Reply:
Hmmm. That sounds like an issue in and of itself.

I think in that case we should go back to the tests you had previously, but without the check on the `exit` code. That's in line with the test coverage of the non-extended version.

Reply:
It's wired and tricky to test a function which called exit() here, maybe a possible way is add a parameter (just like what we already did for driver) to inject the exit(), but it's still a little bit tricky.

Reply:
I dug into this a bit more and found some existing tests that validate the exit behavior. Part of the issue is that this file is for use with `flutter drive`, so the tests need to be run using `flutter drive`. The existing tests are therefore _not_ in the `test` directory.

The code that tests this is in `packages/integration_test/example/test_driver/extended_integration_test.dart`. It's run in `flutter/dev/bots/test.dart#_runWebLongRunningTests`. With that in mind, here's what I'm thinking for validating this new behavior:

* Update `flutter/dev/bots/test.dart#_runFlutterDriverWebTest` to optionally be able to check that a file was written
* Update the call to `_runFlutterDriverWebTest` in `flutter/dev/bots/test.dart#_runWebLongRunninTests` to enable that check on the call for `integration_test.dart` and `extended_integration_test.dart` (around lines 1155 and 1162).

Reply:
Hi @pdblasi-google,

I apologize for the delayed update. Thank you so much for your research and suggestions. I have updated the tests based on your recommendations. Could you please provide me with another review?